compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp

```diff
@@ -0,0 +1,424 @@
+  newShape.insert(newShape.begin(), rhsVals.size());
+  RankedTensorType concatOutType =
+      RankedTensorType::get(newShape, outType.getElementType());
+
+  Value baseOut = base.getDpsInitOperand(0)->get();
+  auto origFill = baseOut.getDefiningOp<linalg::FillOp>();
+  auto origEmpty =
+      origFill.getDpsInitOperand(0)->get().getDefiningOp<tensor::EmptyOp>();
+
+  auto newEmpty = rewriter.create<tensor::EmptyOp>(loc, concatOutType,
+                                                   origEmpty.getDynamicSizes());
+  auto newFill = rewriter.create<linalg::FillOp>(
+      loc, concatOutType, origFill.getDpsInputOperand(0)->get(),
+      newEmpty.getResult());
+
+  Value lhs = base->getOperand(0);
+
+  SmallVector<AffineMap> indexingMaps = base.getIndexingMapsArray();
+  SmallVector<utils::IteratorType> iteratorTypes = base.getIteratorTypesArray();
+  iteratorTypes.insert(iteratorTypes.begin(), utils::IteratorType::parallel);
+
+  indexingMaps[0] = indexingMaps[0].shiftDims(1);
+  indexingMaps[1] = indexingMaps[1].shiftDims(1).insertResult(
+      rewriter.getAffineDimExpr(0), 0);
+  indexingMaps[2] = indexingMaps[2].shiftDims(1).insertResult(
+      rewriter.getAffineDimExpr(0), 0);
+
+  linalg::GenericOp newGenericOp = rewriter.create<linalg::GenericOp>(
+      loc, concatOutType, ValueRange{lhs, newRhs}, newFill.getResult(0),
+      indexingMaps, iteratorTypes);
+  rewriter.cloneRegionBefore(base->getRegion(0), newGenericOp.getRegion(),
+                             newGenericOp.getRegion().begin());
+
+  Value fusedResult = newGenericOp.getResult(0);
+  if (fusionGroup.truncateOps) {
+    // Insert truncate operator.
+    auto truncType =
+        baseTruncOp.value()->getResult(0).getType().cast<RankedTensorType>();
+    auto concatTruncType =
+        RankedTensorType::get(newShape, truncType.getElementType());
+    Value truncOuts = rewriter.create<tensor::EmptyOp>(
+        loc, concatTruncType, origEmpty.getDynamicSizes());
+    SmallVector<AffineMap> truncIndexingMaps(
+        2, AffineMap::getMultiDimIdentityMap(concatTruncType.getRank(),
+                                             rewriter.getContext()));
+    SmallVector<utils::IteratorType> truncIteratorTypes(
+        concatTruncType.getRank(), utils::IteratorType::parallel);
+    auto truncateOpBody = [&](OpBuilder &b, Location loc, ValueRange args) {
+      Value truncOp = b.create<arith::TruncFOp>(
+          loc, concatTruncType.getElementType(), args[0]);
+      rewriter.create<linalg::YieldOp>(loc, truncOp);
+    };
+    auto truncateOp = rewriter.create<linalg::GenericOp>(
+        loc, concatTruncType, newGenericOp->getResults(), truncOuts,
+        truncIndexingMaps, truncIteratorTypes, truncateOpBody);
+
+    fusedResult = truncateOp.getResult(0);
+  }
+
+  SmallVector<Value> newOuts;
+
+  auto fusedResultType = cast<RankedTensorType>(fusedResult.getType());
+  SmallVector<int64_t> shape = llvm::to_vector(fusedResultType.getShape());
```

Comment:
One of SmallVector constructor takes ArrayRef as inputs. so this can be simplified to `SmallVector<int64_t> shape(fusedResultType.getShape());`

Reply:
yeah true... I find this style easier to read.